[ENH] Add advanced-settings docs (#115)
nebari-dev/nebari-docs: docs/docs/explanations/advanced-configuration.md

---

> title: Advanced Configuration

Reviewer: ```suggestion
title: Advanced configuration
```

We've decided to [use sentence case](https://www.nebari.dev/docs/community/style-guide/#capitalization) for headings for consistency :)

```yaml
### Keycloak configuration ###
security:
  keycloak:
    initial_root_password: initpasswd
    overrides:
      image:
        repository: quansight/qhub-keycloak
  ...
```

The `keycloak` section allows you to specify an initial password for the `root` Administrative user to manage your Keycloak database. Which is responsible for managing users, clients, and other Keycloak related configurations. Note that the `root` user is not actually a Nebari user - you cannot access the
main features of Nebari such as JupyterLab with this user. It is exclusively for Keycloak management.

The `overrides` section allows you to specify a custom image for the Keycloak service. This is useful if you want to customize theming or add additional plugins to Keycloak. The full extent of override options can be found in the [Keycloak Helm deployment](https://github.com/codecentric/helm-charts/tree/master/charts/keycloak).

:::warning

> We strongly recommend changing the `initial_root_password` after your initial deployment and deleting this value from your `qhub-config.yaml`. Any changes to this value in the
> `qhub-config.yaml` after the initial deployment will have no effect.

Reviewer: ```suggestion
We strongly recommend changing the `initial_root_password` after your initial deployment and deleting this value from your `nebari-config.yaml`. Any changes to this value in the
`nebari-config.yaml` after the initial deployment will have no effect.
```

Deploying to a local existing kubernetes cluster has different options than the cloud providers. `kube_context` is an optional key that can be used to deploy to a non-default
context. The default node selectors will allow pods to be scheduled anywhere. This can be adjusted to schedule pods on different labeled nodes. Allowing for similar functionality
to node groups in the cloud.

```yaml
### Provider configuration ###
existing:
  kube_context: minikube
  node_selectors:
    general:
      key: kubernetes.io/os
      value: linux
    user:
      key: kubernetes.io/os
      value: linux
    worker:
      key: kubernetes.io/os
      value: linux
```

</TabItem>

<TabItem value="local" label="Local">

Local deployment is intended for Nebari deployments on a "local" cluster created and management by Kind. It is great for experimentation and development.

```yaml
### Provider configuration ###
local:
  kube_context: minikube
  node_selectors:
    general:
      key: kubernetes.io/os
      value: linux
    user:
      key: kubernetes.io/os
      value: linux
    worker:
      key: kubernetes.io/os
      value: linux
```

</TabItem>

</Tabs>

:::note

> Many of the cloud providers regularly update their internal **Kubernetes versions** so if you wish to specify a particular version, please check the following resources. This is _completely optional_ as Nebari will, by default, select the most recent version available for your preferred cloud provider. [Digital Ocean](https://docs.digitalocean.com/products/kubernetes/changelog/) [Google Cloud Platform](https://cloud.google.com/kubernetes-engine/docs/release-notes-stable) [Amazon Web Services](https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html) [Microsoft Azure](https://docs.microsoft.com/en-us/azure/aks/supported-kubernetes-versions?tabs=azure-cli)

Reviewer: ```suggestion
Many of the cloud providers regularly update their internal **Kubernetes versions** so if you wish to specify a particular version, please check the following resources. This is _completely optional_ as Nebari will, by default, select the most recent version available for your preferred cloud provider: [Digital Ocean](https://docs.digitalocean.com/products/kubernetes/changelog/); [Google Cloud Platform](https://cloud.google.com/kubernetes-engine/docs/release-notes-stable); [Amazon Web Services](https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html); [Microsoft Azure](https://docs.microsoft.com/en-us/azure/aks/supported-kubernetes-versions?tabs=azure-cli).
```

name `user-large`. Other cloud providers will have different node labels.

```yaml
### JupyterLab Profiles ###
profiles:
  jupyterlab:
    - display_name: Small Instance
      ...
      kubespawner_override:
        ...
        node_selector:
          "cloud.google.com/gke-nodepool": "user-large"
        ...
```

#### Specifying GPU/Accelerator Requirements

If you want to ensure that you have GPU resources use the following annotations.

```yaml
### JupyterLab Profiles ###
profiles:
  jupyterlab:
    - display_name: Small Instance
      ...
      kubespawner_override:
        ...
        extra_resource_limits:
          nvidia.com/gpu: 1
        ...
```

### Dask Profiles

Finally, we allow for configuration of the Dask workers. In general, similar to the JupyterLab instances you only need to configure the cores and memory.

```yaml
### Dask Profiles ###
profiles:
  ...
  dask_worker:
    "Small Worker":
      worker_cores_limit: 1
      worker_cores: 1
      worker_memory_limit: 1G
      worker_memory: 1G
    "Medium Worker":
      worker_cores_limit: 1.5
      worker_cores: 1.25
      worker_memory_limit: 2G
      worker_memory: 2G
```

> When configuring the memory and CPUs for profiles, there are some important considerations to make. Two important terms to understand are:

Reviewer: I've moved it, feel free to revert if needed. :)

- `limit`: the absolute max memory that a given pod can consume. If a process within the pod consumes more than the `limit` memory the linux OS will kill the process. Limit is not
  used for scheduling purposes with kubernetes.
- `guarantee`: is the amount of memory the kubernetes scheduler uses to place a given pod. In general the `guarantee` will be less than the limit. Often times the node itself has

>   less available memory than the node specification. See this [guide from digital ocean](https://docs.digitalocean.com/products/kubernetes/#allocatable-memory) which is generally

Reviewer: ```suggestion
  less available memory than the node specification. See this [guide from digital ocean](https://docs.digitalocean.com/products/kubernetes/details/limits/#allocatable-memory) which is generally
```

  applicable to other clouds.

For example if a node has 8 GB of ram and 2 CPUs you should guarantee/schedule roughly 75% and follow the digital ocean guide linked above, e.g. 1.5 CPU guaranteed and 5.5 GB
guaranteed.

#### Dask Scheduler

In a few instances, the Dask worker node-group might be running on quite a large instance, perhaps with 8 CPUs and 32 GB of memory (or more). When this is the case, you might also
want to increase the resource levels associated with the Dask Scheduler.

```yaml
### Dask Profiles ###
profiles:
  ...
  dask_worker:
      "Huge Worker":
        worker_cores_limit: 7
        worker_cores: 6
        worker_memory_limit: 30G
        worker_memory: 28G
        scheduler_cores_limit: 7
        scheduler_cores: 6
        scheduler_memory_limit: 30G
        scheduler_memory: 28G
```

## Customizing JupyterHub theme

> Nebari uses a custom JupyterHub theme, [[Quansight/qhub-jupyterhub-theme](https://github.com/quansight/qhub-jupyterhub-theme)](https://github.com/nebari-dev/nebari-jupyterhub-theme). Users can further customize the theme through these available options:

Reviewer: ```suggestion
Nebari uses a custom JupyterHub theme, [nebari-dev/nebari-jupyterhub-theme](https://github.com/nebari-dev/nebari-jupyterhub-theme). Users can further customize the theme through these available options:
```

Below is an example of a custom theme configuration for a GCP deployment:

```yaml
### Theme ###
theme:
  jupyterhub:
    hub_title: demo.nebari.dev
    hub_subtitle: Your open source data science platform, hosted on Google Cloud Platform
    welcome: |
      Welcome! Learn about Nebari's features and configurations in <a href="https://www.nebari.dev/docs">the
      documentation</a>. If you have any questions or feedback, reach the team on
      <a href="https://www.nebari.dev/docs/community#getting-support">Nebari's support
      forums</a>.
    logo: /hub/custom/images/Nebari-Logo-Horizontal-Lockup-White-text.svg
    primary_color: "#cb39ed"
    secondary_color: "#2bd1c5"
    accent_color: "#eda61d"
    text_color: "#1c1d26",
    h1_color: "#0f1015",
    h2_color: "#0f1015",
    navbar_text_color: "#E8E8E8",
    narbar_hover_color: "#00a3b0",
    navbar_color: "#1c1d26",
    display_version: true
    version: v2022.10.1
```

![A demonstration os the theming customizations](/img/how-tos/nebari_login_screen.png)

:::note

> To properly update the image logo, you will must ensure that the provided logo field contains an accessible path to the logo. Which is added in the jupyterhub image, or by passing a valid URL to the logo.

Reviewer: ```suggestion
To properly update the image logo, you must ensure that the provided logo field contains an accessible path to the logo. Which is added in the jupyterhub image, or by passing a valid URL to the logo.
```

- `before_script`: optional script to run before CI starts QHub infrastructure deployment. This is useful in cases that additional setup is required for QHub to deploy the
  resources. Only supported on `gitlab-ci` at the moment.
- `after_script`: optional script to run after CI ends QHub infrastructure deployment. This is useful in cases to notify resources of successful QHub deployment. Only supported on
  `gitlab-ci` at the moment.

If `ci_cd` is not supplied, no CI/CD will be auto-generated, however, we advise employing an infrastructure-as-code approach. This allows teams to more quickly modify their QHub
deployment, empowering developers and data scientists to request the changes and have them approved by an administrator.

### Certificates

`certificate`: This field is used to update the certificate provider used by Nebari. Below we provide the list of supported certificate providers.

<Tabs>
  <TabItem label="Traefik" value="traefik" default="true">

By default, to simplify initial deployment `qhub` uses traefik to create a self-signed certificate. In order to create a certificate that's signed so that web browsers don't throw

> errors we currently support **Let's Encrypt**.

Reviewer: Thanks for the note, Kim. I found this quite confusing too. 

After exploring some docs/issues, I'm gathering the options are: create a new self-signed cert, create a let's encrypt cert, or use an existing self-signed cert. All of these are provided/managed by Traefik. 

I'll update the docs with this, but please feel free to correct me if I'm wrong!

  acme_email: <your-email-address>
  acme_server: https://acme-v02.api.letsencrypt.org/directory
```
:::note
The above snippet will already be present if you provided an `--ssl-cert-email` when you ran `qhub init`.
:::

  </TabItem>
  <TabItem label="Custom" value="Custom">

You may also supply a custom self-signed certificate and secret key.

```yaml
certificate:
  type: existing
  secret_name: <secret-name>
```

To add the tls certificate to kubernetes run the following command with existing files.

```shell
kubectl create secret tls <secret-name> \
  --namespace=<namespace> \
  --cert=path/to/cert/file --key=path/to/key/file
```

:::note
The kubernetes default namespace that QHub uses is `dev`. Otherwise, it will be your `namespace`
defined in the `qhub-config.yaml`.
:::
  </TabItem>
</Tabs>

#### Wildcard certificates

Some of QHub services might require special subdomains under your certificate, Wildcard certificates allow you to secure all subdomains of a domain with a single certificate.
Defining a wildcard certificate decreases the amount of CN names you would need to define under the certificate configuration and reduces the chance of generating a wrong
subdomain.

> It's not possible to request a double wildcard certificate for a domain (for example \*.\*.local.com). As a default behavior of [Traefik](https://doc.traefik.io/traefik/https/tls/#default-certificate), if the Domain Name System (DNS) and Common Name (CN) name doesn't match, Traefik generates and uses a self-signed certificate. This may lead to some unexpected [TLS](https://www.internetsociety.org/deploy360/tls/basics) issues, so as alternative to including each specific domain under the certificate CN list, you may also define a wildcard certificate.

Reviewer: This has been moved to the FAQ section, where the acronyms/definitions work. :)

### Authentication methods

This section walks through security and user authentication as it relates to QHub deployments. There are a few different ways to handle user authentication:

- Auth0
- GitHub
- Password
- Custom OAuth

```yaml
security:
  authentication:

>     type: GitHub

Reviewer: Good catch, yes! Both for "GitHub" and for "Auth0".

[Keycloak as described below](#keycloak).
:::

`security.authentication`: is for configuring the OAuth and GitHub Provider, password based authentication, or custom authentication.

<Tabs>
  <TabItem label="Auth0" value="auth0">

[Auth0](https://auth0.com/#!) can be used for authentication. While it is not free, there is a reasonable free tier that allows deployment of QHub clusters using many different
social providers, passwordless, and email based authentication methods.

QHub has command line options for `qhub init` which automates the creation Auth0 web app via: `--auth-provider=auth0 --auth-auto-provision`.

Otherwise here are docs on [creating an Auth0 Application](https://auth0.com/docs/applications). Make sure to select `Regular Web Application`. Important to note is the
`auth0_subdomain` field which must be only the `<auth0_subdomain>.auth0.com`. So for the following `qhub-dev.auth0.com` the subdomain would be `qhub-dev`. Note that all the
usernames will be the email addresses of users (not usernames).

:::note

> This is a different and distinct step from one outlined in the [Setup] stage.

Reviewer: This section has been re-written. :)

Originally designed for QHub deployments on a "local" minikube cluster, this feature has now expanded to allow users to deploy QHub to any existing kubernetes cluster. The default
options for a `existing` deployment are still set deploy QHub to a minikube cluster.

If you wish to deploy QHub to an existing kubernetes cluster on one of the cloud providers, please refer to a more detailed walkthrough found in the [Deploy QHub to an Existing Kubernetes Cluster].

Deploying to a local existing kubernetes cluster has different options than the cloud providers. `kube_context` is an optional key that can be used to deploy to a non-default
context. The default node selectors will allow pods to be scheduled anywhere. This can be adjusted to schedule pods on different labeled nodes. Allowing for similar functionality
to node groups in the cloud.

```yaml
existing:
  kube_context: minikube
  node_selectors:
    general:
      key: kubernetes.io/os
      value: linux
    user:
      key: kubernetes.io/os
      value: linux
    worker:
      key: kubernetes.io/os
      value: linux
```

</TabItem>

<TabItem value="gcp" label="GCP" default="true" >

> Google Cloud has the best support for QHub and is a great default choice for a production deployment. It allows auto-scaling to zero within the node group. There are no major

Reviewer: Good catch. Updated to say DO is good for testing, but not for production deployment.

`terraform_state` is either `remote`, `existing` or `local` with a default value of `remote`. This decides whether to control the state of the cluster `local` via tfstate file (not
recommended), on an already `existing` terraform state store or remotely and auto creating the terraform state store. See
[terraform remote state](https://www.terraform.io/language/state/remote) docs. If you are doing anything other than testing we highly recommend `remote` unless you know what you
are doing.

The following are examples. `remote` and `local` are straightforward. For a `local` provider that deploys on an existing kubernetes cluster the kubernetes remote backend is used.

```yaml
terraform_state:
  type: remote
```

```yaml
terraform_state:
  type: local
```

Using an existing terraform backend can be done by specifying the `backend` and arbitrary key/value pairs in the `config`.

```yaml
terraform_state:
  type: existing
  backend: s3
  config:
    bucket: mybucket
    key: "path/to/my/key"
    region: "us-east-1"
```

### Default Images

> Default images are to the default image run if not specified in a profile (described in the next section). The `jupyterhub` key controls the jupyterhub image run. These control the

Reviewer: ✅

docker image used to run JupyterHub, the default JupyterLab image, and the default Dask worker image.

```yaml
default_images:
  jupyterhub: "quansight/qhub-jupyterhub:v[QHUB_VERSION]"
  jupyterlab: "quansight/qhub-jupyterlab:v||QHUB_VERSION||"
  dask_worker: "quansight/qhub-dask-worker:v||QHUB_VERSION||"
```

### Storage

Control the amount of storage allocated to shared filesystem.

> NOTE 1: when the storage size is changed, for most providers it will automatically delete (!) the previous storage place. NOTE 2: changing the storage size on an AWS deployment

> > after the initial deployment can be especially tricky so it might be worthwhile padding these storage sizes.

Reviewer: Updated!

```

### JupyterLab Profiles

For each `profiles.jupyterlab` is a named JupyterLab profile.

Use the `kubespawner_override` field to define behavior as per the [KubeSpawner](https://jupyterhub-kubespawner.readthedocs.io/en/latest/spawner.html) API.

It is possible to control which users have access to which JupyterLab profiles. Each profile has a field named `access` which can be set to `all` (default if omitted), `yaml`, or
`keycloak`.

`all` means every user will have access to the profile.

`yaml` means that access is restricted to anyone with their username in the `users` field of the profile or who belongs to a group named in the `groups` field.

`keycloak` means that access is restricted to any user who in Keycloak has either their group(s) or user with the attribute `jupyterlab_profiles` containing this profile name. For
example, if the user is in a Keycloak group named `developers` which has an attribute `jupyterlab_profiles` set to `Large Instance`, they will have access to the Large Instance
profile. To specify multiple profiles for one group (or user) delimit their names using `##` - for example, `Large Instance##Another Instance`.

### Dask Profiles

> Finally, we allow for configuration of the Dask workers. In general, similar to the JupyterLab instances you only need to configuration the cores and memory.

Reviewer 2: ```suggestion
Finally, we allow for configuration of the Dask workers. In general, similar to the JupyterLab instances you only need to configure the cores and memory.
```

> When configuring the memory and CPUs for profiles there are some important considerations to make. Two important terms to understand are:

Reviewer 2: ```suggestion
When configuring the memory and CPUs for profiles, there are some important considerations to make. Two important terms to understand are:
```

> - `limit`: the absolute max memory that a given pod can consume. If a process within the pod consumes more than the `limit` memory the linux OS will kill the process. LimIt is not

Reviewer 2: ```suggestion
- `limit`: the absolute max memory that a given pod can consume. If a process within the pod consumes more than the `limit` memory the linux OS will kill the process. Limit is not
```

  used for scheduling purposes with kubernetes.
- `guarantee`: is the amount of memory the kubernetes scheduler uses to place a given pod. In general the `guarantee` will be less than the limit. Often times the node itself has
  less available memory than the node specification. See this [guide from digital ocean](https://docs.digitalocean.com/products/kubernetes/#allocatable-memory) which is generally
  applicable to other clouds.

> For example if a node has 8 GB of ram and 2 CPUs you should guarantee/schedule roughly 75% and follow the digital ocean guide linked above. For example 1.5 CPU guarantee and 5.5 GB

Reviewer 2: ```suggestion
For example if a node has 8 GB of ram and 2 CPUs you should guarantee/schedule roughly 75% and follow the digital ocean guide linked above, e.g. 1.5 CPU guaranteed and 5.5 GB
```

```yaml
dask_worker:
    "Huge Worker":
      worker_cores_limit: 7
      worker_cores: 6
      worker_memory_limit: 30G
      worker_memory: 28G
      scheduler_cores_limit: 7
      scheduler_cores: 6
      scheduler_memory_limit: 30G
      scheduler_memory: 28G
```

### JupyterLab Profile Node Selectors

A common operation is to target jupyterlab profiles to specific node labels. In order to target a specific node groups add the following. This example shows a GKE node groups with
name `user-large`. Other cloud providers will have different node labels.

```yaml
profiles:
  jupyterlab:
    - display_name: Small Instance
      ...
      kubespawner_override:
        ...
        node_selector:
          "cloud.google.com/gke-nodepool": "user-large"
        ...
```

### Specifying GPU/Accelerator Requirements

If you want to ensure that you have GPU resources use the following annotations.

```yaml
profiles:
  jupyterlab:
    - display_name: Small Instance
      ...
      kubespawner_override:
        ...
        extra_resource_limits:
          nvidia.com/gpu: 1
        ...
```

## Customizing JupyterHub theme

> JupyterHub can be customized since QHub uses [Quansight/qhub-jupyterhub-theme](https://github.com/quansight/qhub-jupyterhub-theme). Available theme options.

Reviewer 2: ```suggestion
Nebari uses a custom JupyterHub theme, [Quansight/qhub-jupyterhub-theme](https://github.com/quansight/qhub-jupyterhub-theme). Users can further customize the theme through these available options:
```

    welcome: |
      Welcome to jupyter.github-actions.qhub.dev. It's maintained by <a href="http://quansight.com">Quansight
      staff</a>. The hub's configuration is stored in a github repository based on
      <a href="https://github.com/Quansight/qhub/">https://github.com/Quansight/qhub/</a>.
      To provide feedback and report any technical problems, please use the
      <a href="https://github.com/Quansight/qhub/issues">github issue tracker</a>.
    logo: /hub/custom/images/jupyter_qhub_logo.svg
    primary_color: '#4f4173'
    secondary_color: '#957da6'
    accent_color: '#32C574'
    text_color: '#111111'
    h1_color: '#652e8e'
    h2_color: '#652e8e'
```

:::note
To properly update the image logo, you will must ensure that the provided logo field contains an accessible **URL** to the logo.
:::

Its also possible to display the current version of qhub by using the `display_version: 'True'` in the above `theme.jupyterhub` configuration. If no extra information is passed,
the displayed version will be the same as `qhub_version`, an overwrite can be done by passing the `version: v.a.b.c` key as well.

## Filesystem environments

> QHub is experimenting with a new way of distributing environments using [conda-store](https://github.com/quansight/conda-store). Please expect this environment distribution method
> to change over time.

Reviewer 2: ```suggestion
Nebari is using a new way of distributing environments using [conda-store](https://github.com/quansight/conda-store). Since conda-store is under active development, be aware that the following details may change over time.
```

Each environment configuration is a `environment.<filename>` mapping to a conda environment definition file. If you need to pin a specific version, please include it in the

> definition. One current requirement is that each environment include `ipykernel`, `ipywidgets`, `qhub-dask==0.2.3`. Upon changing the environment definition expect 1-10 minutes
> upon deployment of the configuration for the environment to appear.

Reviewer 2: ```suggestion
definition. One current requirement is that each environment *must* include `ipykernel`, `ipywidgets`, `qhub-dask==0.2.3`. Upon changing the environment definition expect 1-10 minutes
upon deployment of the configuration for the environment to appear.
```

Reviewer 2: is ` 1-10 minutes` the build time of the environment?

Reviewer: Sounds like it, I'll make a note to update this based on the new conda-store updates.

Nebari comes with two default filesystem environments that are built during deployment.

- filesystem/dask
- filesystem/dashboard

```yaml
environments:
  "environment-default.yaml":
    name: default
    channels:
      - conda-forge
      - defaults
    dependencies:
      - python=3.8
      - ipykernel
      - ipywidgets
      - qhub-dask==||QHUB_VERSION||
      - numpy
      - numba
      - pandas

  "environment-example-2.yaml":
    name: example-2
    channels:
      - conda-forge
      - defaults
    dependencies:
      - python=3.8
      - ipykernel
      - ipywidgets
      - qhub-dask==||QHUB_VERSION||
      - numpy
      - numba
      - pandas
      - jinja2
      - pyyaml
```

:::warning

> By default conda-store restricts the environment channels to only accept `defaults` and `conda-forge`, you can check out [Managing conda environment] for more details.

Reviewer 2: missing link

:::

## Overrides

> Overrides allows you to override the default configuration for a given resource pn Nebari without having to directlu modify the infrastructure components.

Reviewer 2: ```suggestion
Overrides allows you to override the default configuration for a given resource on Nebari without having to directly modify the infrastructure components.
```

Below we show the available resources that can be overridden in the configuration.

<Tabs>

<TabItem label="Jupyterhub" value="jupyterhub" >

JupyterHub uses the [zero to jupyterhub helm chart](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/). This chart has many options that are not configured in the QHub default
installation. You can override specific values in the [values.yaml](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/main/jupyterhub/values.yaml). `jupyterhub.overrides`
is optional.

```yaml
jupyterhub:
  overrides:
    cull:
      users: true
```

</TabItem>

<TabItem label="terraform" value="Terraform">

The QHub configuration file provides a huge number of configuration options for customizing your QHub Infrastructure, while these options are sufficient for an average user, but
aren't exhaustive by any means. There are still a plenty of things you might want to achieve which cannot be configured directly by the above mentioned options, hence we've
introduced a new option called terraform overrides (`terraform_overrides`), which lets you override the values of terraform variables in specific modules/resource. This is a

> relatively advance feature and must be used with utmost care and you should really know, what you're doing.

Reviewer 2: ```suggestion
relatively advance feature and must be used with utmost care and you should really know what you're doing.
```

Here we describe the overrides supported via QHub config file:

### Ingress

You can configure the IP of the load balancer and add annotations for the same via `ingress`'s terraform overrides, one such example for GCP is:

```yaml
ingress:
  terraform_overrides:
      load-balancer-annotations:
          "networking.gke.io/load-balancer-type": "Internal"
          "networking.gke.io/internal-load-balancer-subnet": "pre-existing-subnet"
      load-balancer-ip: "1.2.3.4"
```

This is quite useful for pinning the IP Address of the load balancer.

> ### Deployment inside Virtual Private Network

Reviewer 2: ```suggestion
### Deployment inside a Virtual Private Network
```

#### Azure

> Using terraform overrides you can also deploy inside a virtual private network.

Reviewer 2: ```suggestion
Using terraform overrides you can also deploy inside a virtual private network (VPN).
```

An example configuration for Azure is given below:

```yaml
azure:
  terraform_overrides:
      private_cluster_enabled: true
      vnet_subnet_id: '/subscriptions/<subscription_id>/resourceGroups/<resource_group>/providers/Microsoft.Network/virtualNetworks/<vnet-name>/subnets/<subnet-name>'
  region: Central US
```

#### Google Cloud

> Using terraform overrides you can also deploy inside a VPC in GCP, making the Kubernetes cluster private. Here is an example for configuring the same:

Reviewer 2: is a vpc the same thing as a vpn?

> Using terraform overrides you can also deploy inside a VPC in GCP, making the Kubernetes cluster private. Here is an example for configuring the same:

Reviewer 2: ```suggestion
Using terraform overrides you can also deploy inside a VPC in GCP, making the Kubernetes cluster private. Here is an example configuration:
```

```yaml
google_cloud_platform:
  terraform_overrides:
    networking_mode: "VPC_NATIVE"
    network: "your-vpc-name"
    subnetwork: "your-vpc-subnet-name"
    private_cluster_config:
      enable_private_nodes: true
      enable_private_endpoint: true
      master_ipv4_cidr_block: "172.16.0.32/28"
    master_authorized_networks_config:
      cidr_block: null
      display_name: null
```

> As the name suggests the cluster will be private, which means it would not have access to the internet either, which is not ideal for deploying pods in the cluster, hence we need
> to allow internet access for the cluster, which can be achieved by creating a NAT router by running the following two commands for your vpc network.

Reviewer 2: ```suggestion
As the name suggests the cluster will be private, which means it would not have access to the internet, which is not ideal for deploying pods in the cluster. Therefore, we need
to allow internet access for the cluster, which can be achieved by creating a NAT router by running the following two commands for your VPC network.
```

```
gcloud compute routers create qhub-nat-router --network your-vpc-name --region your-region

gcloud compute routers nats create nat-config --router qhub-nat-router  --nat-all-subnet-ip-ranges --auto-allocate-nat-external-ips --region your-region
```

#### Deployment Notes

Deployment inside a virtual network is slightly different from deploying inside a public network, as the name suggests, since its a virtual private network, you need to be inside

> the network to able to deploy and access QHub. One way to achieve this is by creating a Virtual Machine inside the virtual network, just select the virtual network and subnet name

Reviewer 2: ```suggestion
the network to able to deploy and access QHub. One way to achieve this is by creating a Virtual Machine (VM) inside the virtual network, just select the virtual network and subnet name
```
